src/main/java/io/burt/athena/polling/BackoffPollingStrategy.java

```diff
@@ -13,35 +15,49 @@
     private final Duration maxDelay;
     private final long factor;
     private final Sleeper sleeper;
+    private final Clock clock;
 
     BackoffPollingStrategy(Duration firstDelay, Duration maxDelay) {
-        this(firstDelay, maxDelay, 2L, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()));
+        this(firstDelay, maxDelay, 2L, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()), Clock.systemDefaultZone());
     }
 
     BackoffPollingStrategy(Duration firstDelay, Duration maxDelay, long factor) {
-        this(firstDelay, maxDelay, factor, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()));
+        this(firstDelay, maxDelay, factor, duration -> TimeUnit.MILLISECONDS.sleep(duration.toMillis()), Clock.systemDefaultZone());
     }
 
     BackoffPollingStrategy(Duration firstDelay, Duration maxDelay, Sleeper sleeper) {
-        this(firstDelay, maxDelay, 2L, sleeper);
+        this(firstDelay, maxDelay, 2L, sleeper, Clock.systemDefaultZone());
     }
 
-    BackoffPollingStrategy(Duration firstDelay, Duration maxDelay, long factor, Sleeper sleeper) {
+    BackoffPollingStrategy(Duration firstDelay, Duration maxDelay, Sleeper sleeper, Clock clock) {
+        this(firstDelay, maxDelay, 2L, sleeper, clock);
+    }
+
+    BackoffPollingStrategy(Duration firstDelay, Duration maxDelay, long factor, Sleeper sleeper, Clock clock) {
         this.firstDelay = firstDelay;
         this.maxDelay = maxDelay;
         this.factor = factor;
         this.sleeper = sleeper;
+        this.clock = clock;
     }
 
     @Override
-    public ResultSet pollUntilCompleted(PollingCallback callback) throws SQLException, TimeoutException, ExecutionException, InterruptedException {
+    public ResultSet pollUntilCompleted(PollingCallback callback, Instant deadline) throws SQLException, TimeoutException, ExecutionException, InterruptedException {
         Duration nextDelay = firstDelay;
         while (true) {
-            Optional<ResultSet> resultSet = callback.poll();
+            Optional<ResultSet> resultSet = callback.poll(deadline);
             if (resultSet.isPresent()) {
                 return resultSet.get();
             } else {
-                sleeper.sleep(nextDelay);
+                Duration beforeDeadline = Duration.between(clock.instant(), deadline);
+                if (beforeDeadline.compareTo(nextDelay) < 0) {
+                    if (beforeDeadline.isNegative()) {
+                        throw new TimeoutException();
```

Comment:
Perhaps include an error message here, saying what caused the timeout? Something like "polling reached deadline" (not a stellar message, but it would help differentiate it with `TimeoutException` from network timeouts (or is this completely moot because it will all be wrapped in a `SQLTimeoutException` in the end, with a different message?)

Reply:
Even if it is clouded by `SQLTimeoutException`, I think it makes sense to include an error message anyhow.